### PR TITLE
Rethink the Pipelinesettings API

### PR DESCRIPTION
**Pull request by lailabougria**
- Register should throw if the step id already existed
- Replace should throw if the step id does not exist
- New RegisterOrReplace API is added to support adding or removing, this should never throw
- Remove API is deprecated

### CODE REVIEW

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt [hunk 1]
```diff
@@ -1809,6 +1809,11 @@ namespace NServiceBus.Pipeline
         public void Register<TRegisterStep>()
             where TRegisterStep : NServiceBus.Pipeline.RegisterStep, new () { }
         public void Register(NServiceBus.Pipeline.RegisterStep registration) { }
```
Comment (ramonsmits) [inline src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:1811]: Seeing this makes me wonder why not using `Add`, `AddOrUpdate`, `TryRemove`, `TryUpdate`

which would be more aligned with methods from ConcurrentDictionary https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=netcore-3.1

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 2]
```diff
@@ -22,9 +23,23 @@ public List<RegisterStep> Build()
             var listOfBeforeAndAfterIds = new List<string>();
 
             // Let's do some validation too
+            if (addOrReplaceSteps.Any(addOrReplaceStep => addOrReplaceStep.RegisterStep.StepId != addOrReplaceStep.ReplaceStep.ReplaceId))
+            {
+                throw new Exception("Encountered RegisterOrReplace-registrations in the pipeline for which the ID differs between Add and Replace.");
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:28]: Add the ID's to the exception message?

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 3]
```diff
@@ -45,8 +60,17 @@ public List<RegisterStep> Build()
                 throw new Exception(message);
             }
 
-            //  Step 2: do replacements
-            foreach (var metadata in replacements)
+            //  Step 2: validate and apply replacements
+            var groupedReplacements = replacements.GroupBy(x => x.ReplaceId);
+            if (groupedReplacements.Any(x => x.Count() > 1))
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:65]: Can Count() be .Any()?

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 4]
```diff
@@ -45,8 +60,17 @@ public List<RegisterStep> Build()
...
+            {
+                var duplicateReplaceIdentifiers = groupedReplacements.Where(x => x.Count() > 1).Select(x => x.Key);
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:67]: Same here, Count()>1 to Any() ?

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 5]
```diff
@@ -45,8 +60,17 @@ public List<RegisterStep> Build()
...
+                var duplicateIdentifiersList = string.Join(",", duplicateReplaceIdentifiers);
+                var message = $"You can only replace an existing step once, '{duplicateIdentifiersList}' were detected more than once in the pipeline replacement process";
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:69]: I personally don't like lists in the middle of an exception message but rather see these appended to the exception message but that is personal preference.

File: src/NServiceBus.Core/Pipeline/PipelineSettings.cs [hunk 6]
```diff
@@ -76,11 +66,62 @@ public void Replace<T>(string stepId, Func<IServiceProvider, T> factoryMethod, s
...
+        /// <summary>
+        /// Registers a new step behavior or replaces the existing one with the same id with a new one.
+        /// </summary>
+        /// <param name="stepId">The identifier of the step to replace its implementation.</param>
+        /// <param name="factoryMethod">The factory method to create new instances of the behavior.</param>
+        /// <param name="description">The description of the new behavior.</param>
+        public void RegisterOrReplace<T>(string stepId, Func<IServiceProvider, T> factoryMethod, string description = null)
+            where T : IBehavior
+        {
+            BehaviorTypeChecker.ThrowIfInvalid(typeof(T), "behavior");
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineSettings.cs:111]: Shouldn't that be `nameof(T)`? Not sure if that works.

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt (hunk 1 above)
Comment (lailabougria) [reply src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:1811]: That would imply renaming our current `Register` to `Add`, breaking user-code unnecessarily. The initial name I added was `AddOrReplace` and I renamed it to `RegisterOrReplace` to remove the discrepancy in the added API, instead of breaking the existing one.

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 3 above)
Comment (lailabougria) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:65]: If you mean:
if (groupedReplacements.Any(x => x.Any()))
then yes, but is it more readable? I'm on the fence

File: src/NServiceBus.Core.Tests/Pipeline/BehaviorRegistrationsCoordinatorTests.cs [hunk 7]
```diff
@@ -69,6 +71,32 @@ public void Registrations_Replace()
             Assert.AreEqual("2", model[1].Description);
         }
 
+        [Test]
+        public void Registrations_AddOrReplace_WhenDoesNotExist()
+        {
+            addOrReplacements.Add( AddOrReplaceStep.Create("1", typeof(ReplacedBehavior), "new"));
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Pipeline/BehaviorRegistrationsCoordinatorTests.cs:77]: Nit: Remove blankspace

File: src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs [hunk 8]
```diff
@@ -34,14 +34,63 @@ public void ShouldOnlyAllowReplacementOfExistingRegistrations()
...
+            },
+            new List<AddOrReplaceStep>());
 
             var ex = Assert.Throws<Exception>(() => builder.Build());
 
             Assert.AreEqual("You can only replace an existing step registration, 'DoesNotExist' registration does not exist.", ex.Message);
         }
 
+        [Test]
+        public void ShouldAddWhenAddingOrReplacingABehaviorThatDoesntYetExist()
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs:46]: Nit: think you an drop the "Yet" in the name since it doesn't add anything?

File: src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs [hunk 9]
```diff
@@ -34,14 +34,63 @@ public void ShouldOnlyAllowReplacementOfExistingRegistrations()
...
+
+            var model = builder.Build();
+
+            Assert.That(model.Count, Is.EqualTo(2));
+            var overriddenBehavior = model.FirstOrDefault(x => x.StepId == "SomeBehaviorOfParentContext");
+            Assert.That(overriddenBehavior, Is.Not.Null);
+            Assert.That(overriddenBehavior.BehaviorType, Is.EqualTo(typeof(AnotherBehaviorOfParentContext)));
+        }
+
+
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs:93]: Nit: remove extra line

File: src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs [hunk 10]
```diff
@@ -109,8 +163,7 @@ public void ShouldDetectMissingBehaviorForRootContext()
             {
                 RegisterStep.Create("Child", typeof(ChildBehaviorOfChildContext), "desc"),
 
-            }, new List<RemoveStep>(), new List<ReplaceStep>());
-
+            }, new List<RemoveStep>(), new List<ReplaceStep>(), new List<AddOrReplaceStep>());
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Pipeline/PipelineModelBuilderTests.cs:166]: I wonder if we could add some test helper method to create the PipelineModelBuilder to be able to skip newing up all the empty lists? (for readability purposes)

File: src/NServiceBus.Core/Pipeline/AddOrReplaceStep.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+namespace NServiceBus
+{
+    using System;
+    using Pipeline;
+
+    class AddOrReplaceStep
+    {
+        private AddOrReplaceStep(string stepId, RegisterStep registerStep, ReplaceStep replaceStep)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/AddOrReplaceStep.cs:8]: Nit: Skip the `private`

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 12]
```diff
@@ -22,9 +23,23 @@ public List<RegisterStep> Build()
             var listOfBeforeAndAfterIds = new List<string>();
 
             // Let's do some validation too
+            if (addOrReplaceSteps.Any(addOrReplaceStep => addOrReplaceStep.RegisterStep.StepId != addOrReplaceStep.ReplaceStep.ReplaceId))
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:26]: Looking at the code in AddOrReplaceStep.cs can this really happen?

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 5 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:69]: I like to lead with the problem detected and I also think the word process can be misleading the user, how about

>Multiple replacements of the same pipeline behaviour is not supported. Make sure that you only register a single replacement for:
> - ID1
> - ID2

**Review by andreasohlund — COMMENTED** [review]
Mostly nitpicks, approach looks good IMO

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 12 above)
Comment (lailabougria) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:26]: Not anymore, thanks for pointing that out. It could happen but I refactored the AddOrReplaceStep creation

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 2 above)
Comment (lailabougria) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:28]: This is dropped as it can't occur anymore

File: src/NServiceBus.Core/Pipeline/PipelineSettings.cs (hunk 6 above)
Comment (lailabougria) [reply src/NServiceBus.Core/Pipeline/PipelineSettings.cs:111]: The `ThrowIfInvalid`-method expects a Type parameter, so `nameof(T)` wouldn't work as it would return the type name

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 3 above)
Comment (lailabougria) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:58]: Looking at the code, I take back what I said. ´if (groupedReplacements.Any(x => x.Any()))´ wouldn't work, because what we're testing here, is to verify if there are multiple replacements. So we're not checking if there are `Any`, but actually "more than one".

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt (hunk 1 above)
Comment (ramonsmits) [reply src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:1811]: True about breaking, but I'm not expecting our users to have lots of lines of pipeline configuration. Also, that is the beauty of our obsolete practice that the compiler errors give guidance to use a different method and the fix is super simple.

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt [hunk 13]
```diff
@@ -1809,6 +1809,12 @@ namespace NServiceBus.Pipeline
...
+        public void RegisterOrReplace(string stepId, System.Type behavior, string description = null) { }
+        public void RegisterOrReplace<T>(string stepId, T behavior, string description = null)
+            where T : NServiceBus.Pipeline.IBehavior { }
+        public void RegisterOrReplace<T>(string stepId, System.Func<System.IServiceProvider, T> factoryMethod, string description = null)
+            where T : NServiceBus.Pipeline.IBehavior { }
+        [System.ObsoleteAttribute(@"Removing behaviors from the pipeline is discouraged, to disable a behavior replace the behavior by an empty one. Documentation: https://docs.particular.net/nservicebus/pipeline/manipulate-with-behaviors. The member currently throws a NotImplementedException. Will be removed in version 9.0.0.", true)]
```
Comment (ramonsmits) [inline src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:1817]: This should be moved to a separate file. https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/obsoletes-v8.cs

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs (hunk 3 above)
Comment (ramonsmits) [reply src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:58]: Sorry, I thought it was `>=1` my bad!

File: src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs [hunk 14]
```diff
@@ -8,23 +8,31 @@ namespace NServiceBus
...
+            this.addOrReplaceSteps = addOrReplaceSteps;
         }
 
         public List<RegisterStep> Build()
         {
             var registrations = new Dictionary<string, RegisterStep>(StringComparer.CurrentCultureIgnoreCase);
             var listOfBeforeAndAfterIds = new List<string>();
 
-            // Let's do some validation too
+            var totalAdditions = addOrReplaceSteps.Where(addOrReplaceStep => additions.All(addition => addition.StepId != addOrReplaceStep.StepId))
```
Comment (ramonsmits) [inline src/NServiceBus.Core/Pipeline/PipelineModelBuilder.cs:24]: This seem to be set operations, maybe possible to use `Enumerable.Except`?

File: src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt (hunk 13 above)
Comment (lailabougria) [reply src/NServiceBus.Core.Tests/ApprovalFiles/APIApprovals.ApproveNServiceBus.approved.txt:1817]: But this is the approval file? The obsoleted method is in the file you specified.

**Review by ramonsmits — APPROVED** [review]
